modules/core/src/smithy4s/internals/SchematicDocumentDecoder.scala

```diff
@@ -344,49 +345,103 @@ object SchematicDocumentDecoder
     }
   }
 
-  def union[S](
-      first: Alt[DocumentDecoderMake, S, _],
-      rest: Vector[Alt[DocumentDecoderMake, S, _]]
-  )(
-      total: S => Alt.WithValue[DocumentDecoderMake, S, _]
-  ): DocumentDecoderMake[S] = Hinted.static {
+  private def handleUnion[S](
+      f: (List[PayloadPath.Segment], Document) => S
+  ): DocumentDecoder[S] =
     new DocumentDecoder[S] {
+      def expected: String = "Union"
+      def apply(pp: List[PayloadPath.Segment], document: Document): S =
+        f(pp, document)
+      def canBeKey: Boolean = false
+    }
 
-      def jsonLabel[A](alt: Alt[DocumentDecoderMake, S, A]): String =
-        alt.instance.hints.get(JsonName).map(_.value).getOrElse(alt.label)
-
-      val decoders: Map[String, (List[PayloadPath.Segment], Document) => S] =
-        (first +: rest).map { case alt @ Alt(_, instance, inject) =>
-          val encoder = { (pp: List[PayloadPath.Segment], doc: Document) =>
-            inject(instance.get.apply(jsonLabel(alt) :: pp, doc))
+  private type DecoderMap[S] =
+    Map[String, (List[PayloadPath.Segment], Document) => S]
+
+  private def discriminatedUnion[S](
+      discriminated: Discriminated,
+      decoders: DecoderMap[S]
+  ): DocumentDecoder[S] = handleUnion {
+    (pp: List[PayloadPath.Segment], document: Document) =>
+      document match {
+        case DObject(map) if (map.size > 1) =>
+          map
+            .get(discriminated.propertyName) match {
+            case Some(value: Document.DString) =>
+              decoders.get(value.value) match {
+                case Some(decoder) => decoder(pp, document)
+                case None =>
+                  throw new PayloadError(
+                    PayloadPath(pp.reverse),
+                    "Union",
+                    s"Unknown discriminator: ${value.value}"
+                  )
+              }
+            case _ =>
+              throw new PayloadError(
+                PayloadPath(pp.reverse),
+                "Union",
+                s"Unable to locate discriminator under property '${discriminated.propertyName}'"
+              )
           }
-          jsonLabel(alt) -> encoder
-        }.toMap
+        case _ =>
+          throw new PayloadError(
+            PayloadPath(pp.reverse),
+            "Union",
+            "Expected more than single-key Json object"
```

Comment:
Not necessarily : the alternative could be an empty structure

Reply:
Ah yeah good call, I will update and add a test case to make sure that is handled